# Домашнє завдання до теми "Розширені методи обробки даних "

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [ ]:
import pandas as pd
df = pd.read_csv('supermarket_sales.csv')

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [ ]:
df.sort_values('Unit price', ascending=False)[:5]

Invoice ID Branch       City Customer type Gender         Product line  \
983  148-41-7930      C  Naypyitaw        Normal   Male    Health and beauty   
122  219-22-9386      B   Mandalay        Member   Male    Sports and travel   
930  641-62-7288      B   Mandalay        Normal   Male   Home and lifestyle   
494  437-53-3084      B   Mandalay        Normal   Male  Fashion accessories   
283  667-92-0055      A     Yangon        Member   Male    Health and beauty   

     Unit price  Quantity  Tax 5%    Total       Date   Time      Payment  \
983       99.96         7  34.986  734.706  1/23/2019  10:33         Cash   
122       99.96         9  44.982  944.622   3/9/2019  17:26  Credit card   
930       99.92         6  29.976  629.496  3/24/2019  13:33      Ewallet   
494       99.89         2   9.989  209.769  2/26/2019  11:48      Ewallet   
283       99.83         6  29.949  628.929   3/4/2019  15:02      Ewallet   

       cogs  gross margin percentage  gross income  Rating Rating Category  \
983  699.72                 4.761905        34.986     6.1          Medium   
122  899.64                 4.761905        44.982     4.2             Low   
930  599.52                 4.761905        29.976     7.1          Medium   
494  199.78                 4.761905         9.989     7.1          Medium   
283  598.98                 4.761905        29.949     8.5            High   

     Hour  
983    10  
122    17  
930    13  
494    11  
283    15

2. Знайдіть інвойс (рядок в даних) в категорії товарів Electronic accessories, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

Варіант 1. Сортування спершу за Quantity

In [ ]:
df[df['Product line']=='Electronic accessories'].sort_values(by=["Quantity", "Unit price"], ascending=[False, False]).head(1)

Invoice ID Branch    City Customer type Gender            Product line  \
105  704-48-3927      A  Yangon        Member   Male  Electronic accessories   

     Unit price  Quantity  Tax 5%    Total       Date   Time  Payment   cogs  \
105       88.67        10  44.335  931.035  1/12/2019  14:50  Ewallet  886.7   

     gross margin percentage  gross income  Rating  
105                 4.761905        44.335     7.3

Варіант 2. Сортування спершу за Unit pricy

In [ ]:
df[df['Product line']=='Electronic accessories'].sort_values(by=['Unit price', 'Quantity'], ascending=[False, False])[:1]

Invoice ID Branch      City Customer type  Gender  \
209  817-69-8206      B  Mandalay        Normal  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
209  Electronic accessories       99.73         9  44.8785  942.4485   

         Date   Time      Payment    cogs  gross margin percentage  \
209  3/2/2019  19:42  Credit card  897.57                 4.761905   

     gross income  Rating  
209       44.8785     6.5

3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу: високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 до 8 не включаючи - то він середній (`Medium`)
- інакше - низький (`Low`).

In [ ]:
def get_rating_category(val):
  if val >= 8:
    return "High"
  elif val >= 6 and val < 8:
    return "Medium"
  else:
    return "Low"

In [ ]:
df['Rating Category'] = df.Rating.apply(get_rating_category)

3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [ ]:
df['Rating Category'].value_counts(normalize=True)

Rating Category
Medium    0.356
High      0.329
Low       0.315
Name: proportion, dtype: float64

4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [ ]:
time = '11:28'

In [ ]:
# option 1

def get_hour(time):
    return int(time.split(':')[0])

# option 2
def get_hour(time):
    return int(time[:2])

In [ ]:
get_hour(time)

11

4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [ ]:
df['Hour'] = df.Time.apply(get_hour)

4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [ ]:
df.Hour.value_counts()

Hour
19    113
13    103
15    102
10    101
18     93
11     90
12     89
14     83
16     77
20     75
17     74
Name: count, dtype: int64

Альтернативно:

In [ ]:
df.groupby('Hour').size()

Hour
10    101
11     90
12     89
13    103
14     83
15    102
16     77
17     74
18     93
19    113
20     75
dtype: int64

Найбільше покупок відбувається о 19:00 — 113 разів. Це може бути пов'язано з тим, що багато людей здійснюють покупки після роботи чи навчання, ввечері плануючи вечерю або інші домашні потреби. Також це може бути час, коли супермаркети пропонують спеціальні знижки або акції на кінець дня, щоб збільшити продажі.

В години обідньої перерви, о 13:00, також спостерігається велика кількість покупок — 103. Це може свідчити про те, що люди використовують свою перерву на обід, щоб зробити швидкі покупки або купити щось необхідне в середині робочого дня.

Відносно високі показники покупок о 15:00 — 102 покупки — можуть бути пов'язані з завершенням робочого дня для деяких людей, або з початком знижок та акцій, які часто стартують у другій половині дня.

Що стосується 10:00 ранку з кількістю покупок 101, це може бути пік ранкових покупок, коли люди, можливо, купують продукти на день або виконують планові покупки, оскільки багато супермаркетів починають роботу о цій годині. Це також час, коли супермаркети можуть бути менш забиті, що дозволяє швидше та комфортніше робити покупки.

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [ ]:
df.groupby('City')['Total'].mean()

City
Mandalay     319.872506
Naypyitaw    337.099715
Yangon       312.354031
Name: Total, dtype: float64

6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [ ]:
df.groupby('City')['Total'].agg(['count', 'mean', 'min', 'max'])

count        mean      min      max
City                                          
Mandalay     332  319.872506  18.6375  1022.49
Naypyitaw    328  337.099715  10.6785  1042.65
Yangon       340  312.354031  12.6945  1039.29

**Висновки:**

1. **Naypyitaw має найвищу середню суму чеку** (337.10), що може вказувати на більш високу купівельну спроможність або на те, що там продаються дорожчі товари. На це вказує і той факт, що тут найвищий максимальний чек.
2. **Mandalay має середню суму чеку**, схожу на Yangon (319.87), але із нижчим максимальним чеком (1022.49), що може вказувати на більш однорідні покупки.

**Рекомендації:**

1. **Naypyitaw**:
   - Оскільки Naypyitaw має найвищу середню і максимальну суму чеків, маркетингові стратегії можуть бути спрямовані на преміальний сегмент ринку. Пропозиція ексклюзивних товарів або брендів, які приваблюють клієнтів з вищою купівельною спроможністю, може ще більше підвищити продажі.
   - Запровадження спеціальних подій чи заходів, що пропонують унікальний досвід покупок, може допомогти зберегти існуючих клієнтів і залучити нових.
   - Розробка програм лояльності, які нагороджують клієнтів за частіші та більші покупки, може заохотити збільшення середнього чеку.

2. **Mandalay**:
   - З огляду на нижчий максимальний чек, супермаркет у Mandalay може зосередитись на пропозиціях, які підкреслюють цінність і доступність, щоб привабити більше клієнтів, які здійснюють регулярні покупки.
   - Проведення акцій і знижок може стимулювати клієнтів купувати більше за один візит, збільшуючи середню суму чеку.
   - Вивчення попиту на певні категорії товарів і адаптація асортименту також може допомогти підвищити суму середнього чеку.

**Додаткові рекомендації:**

- **Для обох міст**: Важливо провести детальний аналіз покупців, щоб зрозуміти їхні переваги та поведінку. Використання CRM(Customer Relationship Management)-систем може допомогти зібрати та аналізувати дані про клієнтів, щоб створювати персоналізовані пропозиції.
- **Оцінка запасів та управління ними**: Ефективне управління запасами може допомогти забезпечити наявність популярних товарів і своєчасне видалення непопулярних, щоб максимізувати прибуток і мінімізувати втрати.
- **Аналіз конкурентів**: Розуміння того, як конкуренти впливають на продажі, може допомогти визначити унікальні торгові пропозиції та виробити конкурентні стратегії ціноутворення.

7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [ ]:
df.to_csv('../data/supermarket_sales_enriched.csv', index=False)